### ISSUE TITLE
Why is the continous_loader replacing components instead of updating existing ones?

### ISSUE THREAD

**Issue by Unarelith** [skypjack/entt#476]
https://gist.github.com/Unarelith/5b8affa38a3eddf2a03647c218ea80c5

In that case, I'm serializing `ComponentA` and `ComponentB` with the snapshot, and I'm deserializing into `ComponentA2` and `ComponentB`

I load the archive once, [set `fvalue`](https://gist.github.com/Unarelith/5b8affa38a3eddf2a03647c218ea80c5#file-main-cpp-L124) of `ComponentA2` to a specific value before reloading the archive, and that reload resets `fvalue`

So my guess is that the `continous_loader` is replacing components instead of updating them, but why is it replacing them?

**Comment by skypjack** [comment]
The `continuous_loader` is quite simple and somehow _stupid_. It's originally meant to keep in sync a local entity with its remote representation, so the remote one will always take precedence.
Consider that a lot of things I develop for other purposes (or real world software) enter `EnTT` (and sometimes exit it). So, the _reasons_ behind something could be strictly related to a particular problem I faced.

Roughly speaking, in your case it's as if the remote entity were mirrored locally. No matter what you do, the update you receive is applied and overwrites the local changes.
When you reload the archive, it behaves as if it received an update from a remote system and therefore it considers these data _more recent_ than the local representation.

>So my guess is that the continous_loader is replacing components instead of updating them, but why is it replacing them?

What's the difference exactly?
I receive a representation of `T` for entity `e` and have two options:
* Replace it: all data are wiped out and the new representation is used instead.
* Update it... how exactly? I've a whole object, for what is worth this is the most recent representation of which I know and I've no idea if you changed locally the data member `x` or `y`, so replacing the component as a whole is still the only possible solution.

I've no idea what your use case is but probably what you are looking for **is not** the continuous loader. :)

**Comment by Unarelith** [comment]
> What's the difference exactly?
> I receive a representation of `T` for entity `e` and have two options:
> 
> * Replace it: all data are wiped out and the new representation is used instead.
> * Update it... how exactly? I've a whole object, for what is worth this is the most recent representation of which I know and I've no idea if you changed locally the data member `x` or `y`, so replacing the component as a whole is still the only possible solution.

Why should you know the local changes? I think that the continous loader could actually try to get the component from the registry if it exists (or create it if it doesn't exist), and pass it to the loader archive, instead of creating a new one.

That way, the archive itself can manage how the data will be loaded and if it will overwrite existing data, instead of having the continous loader replacing the entire component.

Maybe it's only suitable for my test case though?

> I've no idea what your use case is but probably what you are looking for **is not** the continuous loader. :)

Maybe, actually I'm quite new to using ECS with networking, so I'm quite confused on the right way to go. The whole point of this is to add entities to [my project](https://github.com/Unarelith/OpenMiner).

I wanted to be sure that keeping the client registry and the server registry in sync for some components would be possible. But apparently it's not, so I guess I'll have to find another way.. :/

**Comment by skypjack** [comment]
>Why should you know the local changes? I think that the continous loader could actually try to get the component from the registry if it exists (or create it if it doesn't exist), and pass it to the loader archive, instead of creating a new one.

The archive is used to read the data from whatever form your packet has.
Once the component is read, the loader uses it to create or replace the local instance.
What would be the point of passing it back to the loader that just returned the component itself?

>The whole point of this is to add entities to my project.

Wow. I like your project! Even though I'm biased biased towards Minecraft probably. :smile: 

>I wanted to be sure that keeping the client registry and the server registry in sync for some components would be possible. But apparently it's not, so I guess I'll have to find another way.. :/

It's possible, of course. I wouldn't use the snapshot stuff though.
Long story short: that's the result of a work done on an older version of `EnTT` and nowadays the whole registry offers an API that makes all this easier to achieve.
Btw the continuous loader is meant to load a stream of data from a source directly into the registry. If a message arrives with an updated version of the component `C` for entity `e`, then it slaps the former on whatever is assigned to the latter. In case this isn't the desired behavior, that loader isn't the right solution for your problem.

**Comment by Unarelith** [comment]
> The archive is used to read the data from whatever form your packet has.
> Once the component is read, the loader uses it to create or replace the local instance.
> What would be the point of passing it back to the loader that just returned the component itself?

I mean, instead of deserializing the packet into a newly created component that will replace the local instance, the archive could instead use the local instance and work directly on it.

> Wow. I like your project! Even though I'm biased biased towards Minecraft probably. 

Thanks a lot! :D 
It's still missing a lot of features but in a few years it may finally become usable. :p

> It's possible, of course. I wouldn't use the snapshot stuff though.

Ok so now I'm trying to think of a good way without the continous loader.

I came up with something like this:
```cpp
void NetworkController::update(entt::registry &registry) {
	registry.view<NetworkComponent, PositionComponent>().each([this] (auto, auto &network, auto &position) {
		if (position.isUpdated)
			m_server->sendEntityPosUpdate(network.entityID, position.x, position.y, position.z);
	});

	registry.view<NetworkComponent, DrawableDef>().each([this] (auto, auto &network, auto &drawableDef) {
		if (drawableDef.isUpdated)
			m_server->sendEntityDrawableDef(network.entityID, drawableDef);
	});
}

void NetworkController::sendEntities(entt::registry &registry, const ClientInfo &client) {
	registry.view<NetworkComponent>().each([&] (auto entity, auto &network) {
		m_server->sendEntitySpawn(network.entityID, &client);

		if (auto *position = registry.try_get<PositionComponent>(entity) ; position) {
			m_server->sendEntityPosUpdate(network.entityID, position->x, position->y, position->z);
		}

		if (auto *drawableDef = registry.try_get<DrawableDef>(entity) ; drawableDef) {
			m_server->sendEntityDrawableDef(network.entityID, *drawableDef);
		}
	});
}
```

The `update` function is called every main loop turn, and the `sendEntities` functions is sent to a new client when it connects to the server. The `m_server->sendXXX` functions are used to send the actual packets to the client.

Is there better ways to do what I did in that code regarding `entt`?

**Comment by skypjack** [comment]
>I mean, instead of deserializing the packet into a newly created component that will replace the local instance, the archive could instead use the local instance and work directly on it.

The fact is that there is no way to know what members to pick from the component being serialized.
So, even in this case it would get the local instance and overwrite all its member anyway.

>Is there better ways to do what I did in that code regarding entt?

First of all, I'd use a dedicate tag component rather than doing something like this:
```
if (position.isUpdated)
```
Similarly for all other `if`s. This way, you can turn this:
```
registry.view<NetworkComponent, PositionComponent>().each([this] (auto, auto &network, auto &position) {
    if (position.isUpdated)
        m_server->sendEntityPosUpdate(network.entityID, position.x, position.y, position.z);
    });
```
Into this:
```
auto data = view.<NetworkComponent, PositionComponent>();
for(auto entity: registry.view<HasUpdatedPosition>()) {
    auto &&[network, position] = data.get<Network, Position>(entity);
    m_server->sendEntityPosUpdate(network.entityID, position.x, position.y, position.z);
}
```
Or something similar. Just thinking aloud.

**Comment by Unarelith** [comment]
> The fact is that there is no way to know what members to pick from the component being serialized.
> So, even in this case it would get the local instance and overwrite all its member anyway.

Not when you don't serialize/deserialize all the fields of a component. If you want local component members in both registries, the continous_loader currently doesn't allow it.

However, for most use cases it's not a problem since you're overwriting all the members anyway. In my use case it was, because I was keeping local client-only data in the component. :/

> First of all, I'd use a dedicate tag component rather than doing something like this:
> 
> ```
> if (position.isUpdated)
> ```
> 
> Similarly for all other `if`s. This way, you can turn this:
> 
> ```
> registry.view<NetworkComponent, PositionComponent>().each([this] (auto, auto &network, auto &position) {
>     if (position.isUpdated)
>         m_server->sendEntityPosUpdate(network.entityID, position.x, position.y, position.z);
>     });
> ```
> 
> Into this:
> 
> ```
> auto data = view.<NetworkComponent, PositionComponent>();
> for(auto entity: registry.view<HasUpdatedPosition>()) {
>     auto &&[network, position] = data.get<Network, Position>(entity);
>     m_server->sendEntityPosUpdate(network.entityID, position.x, position.y, position.z);
> }
> ```
> 
> Or something similar. Just thinking aloud.

Hmm interesting, thanks, I'll look into that.
But what is a tag component exactly? An empty struct?

**Comment by skypjack** [comment]
>because I was keeping local client-only data in the component.

Ahaaaaaa... ok, I missed this point. Yeah, it won't work in this case since it will keep overwriting your local data. Either split the component in two parts or just go straight to the point with the API of the registry. I'd lean towards the second option though.

>But what is a tag component exactly? An empty struct?

Yup, sorry. An empty type. They are also optimized and never allocated for what is worth.

**Comment by Unarelith** [comment]
> Ahaaaaaa... ok, I missed this point. Yeah, it won't work in this case since it will keep overwriting your local data. Either split the component in two parts or just go straight to the point with the API of the registry. I'd lean towards the second option though.

That's why I said that it could've been nice to not overwrite components if a local copy exists. :p
IMO the archive itself should decide if it overwrites all the fields or not.

Anyway, I'm not using the continous loader anymore so it's all good

> Yup, sorry. An empty type. They are also optimized and never allocated for what is worth.

Ohhh ok. Thanks! I've a last question: What's the cost of creating a view? Given your code, I assume that creating a view is trivial but using it is not, am I right?

**Comment by skypjack** [comment]
Creating a view is definitely cheap. Using it... it depends, mainly on the number of entities to which the components are assigned. I think this is expected though, so either the question is somehow odd or I didn't get what you mean.

**Comment by Unarelith** [comment]
Ok yeah I guess that was a stupid question. :')

Thanks a lot for your answers!
